kotlinx-coroutines-core/jvm/test/ThreadContextMutableCopiesTest.kt

```diff
@@ -0,0 +1,113 @@
+ */
+
+package kotlinx.coroutines
+
+import kotlin.coroutines.*
+import kotlin.test.*
+
+class ThreadContextMutableCopiesTest : TestBase() {
+    companion object {
+        val threadLocalData: ThreadLocal<MutableList<String>> = ThreadLocal.withInitial { ArrayList() }
+    }
+
+    class MyMutableElement(
+        private val mutableData: MutableList<String>
+    ) : CopyableThreadContextElement<MutableList<String>, MyMutableElement> {
+
+        companion object Key : CoroutineContext.Key<MyMutableElement>
+
+        override val key: CoroutineContext.Key<*>
+            get() = Key
+
+        override fun updateThreadContext(context: CoroutineContext): MutableList<String> {
+            val st = threadLocalData.get()
+            threadLocalData.set(mutableData)
+            return st
+        }
+
+        override fun restoreThreadContext(context: CoroutineContext, oldState: MutableList<String>) {
+            threadLocalData.set(oldState)
+        }
+
+        override fun copyForChildCoroutine(): MyMutableElement {
+            return MyMutableElement(ArrayList(mutableData))
+        }
+
+        override fun merge(element: MyMutableElement): MyMutableElement {
+            return MyMutableElement((element.mutableData + mutableData).toSet().toMutableList())
+        }
+    }
+
+    @Test
+    fun testDataIsCopied() = runTest {
+        val root = MyMutableElement(ArrayList())
+        runBlocking(root) {
+            val data = threadLocalData.get()
+            expect(1)
+            launch(root) {
+                assertNotSame(data, threadLocalData.get())
+                assertEquals(data, threadLocalData.get())
+                finish(2)
+            }
+        }
+    }
+
+    @Test
+    fun testDataIsNotOverwritten() = runTest {
+        val root = MyMutableElement(ArrayList())
+        runBlocking(root) {
+            expect(1)
+            val originalData = threadLocalData.get()
+            threadLocalData.get().add("X")
+            launch {
+                threadLocalData.get().add("Y")
+                // Note here, +root overwrites the data
+                launch(Dispatchers.Default + root) {
+                    assertEquals(listOf("X", "Y"), threadLocalData.get().sorted())
+                    assertNotSame(originalData, threadLocalData.get())
+                    finish(2)
+                }
+            }
+        }
+    }
+
+    @Test
+    fun testDataIsMerged() = runTest {
+        val root = MyMutableElement(ArrayList())
+        runBlocking(root) {
+            expect(1)
+            val originalData = threadLocalData.get()
+            threadLocalData.get().add("X")
+            launch {
+                threadLocalData.get().add("Y")
+                // Note here, +root overwrites the data
+                launch(Dispatchers.Default + MyMutableElement(mutableListOf("Z"))) {
+                    assertEquals(listOf("X", "Y", "Z"), threadLocalData.get().sorted())
```

Comment:
>I think the unit tests require sorted() because runBlocking {} is a corner case?

It requires `sorted()` just because I wrote `merge` function this way. I've rewritten it and it no longer requires `sorted()`

>merge() is also a member function. So, if the left-handed and right-handed values were different subclasses of the same supertype that so-happened to use the same Element key, a different merge() function might be called depending on whether or not the left-handed input was present.

I'm not sure that we want to actively promote use-case when different types of elements have the same key. 
If so, then it's the sole responsibility of `merge` function to take care of the polymorphic typing.

>If in most cases, Coroutines were to call leftHanded.merge(rightHanded) but sometimes it were to call rightHanded.merge(EmptyCoroutineContext), the implementation of merge() would have to branch and introspect to figure out how it was called. 


The only guarantee `merge` provides is that this member will be invoked on an element if it is being overwritten by another element with the same key during a new coroutine creation.